<<cache=FALSE, message=FALSE, warning=FALSE>>=
source('~/src/git/cadmium/src/report/R/classification.R')
source('~/src/git/cadmium/src/report/R/loadData.R')
@

<<cache=TRUE, message=FALSE, warning=FALSE>>=
query <- LoadQuery(
  aafile="~/src/git/cadmium/input/faa/Arabidopsis_thaliana.faa",
  gfffile="~/src/git/cadmium/input/gff/Arabidopsis_thaliana.gff",
  orphanfile="~/src/git/cadmium/input/orphan-list.txt"
)
target <- LoadTarget(
  aafile="~/src/git/cadmium/input/faa/Arabidopsis_lyrata.faa",
  dnafile="~/src/git/cadmium/input/fna/Arabidopsis_lyrata.fna",
  sifile="~/src/git/cadmium/input/maps/Arabidopsis_thaliana.vs.Arabidopsis_lyrata.map.tab",
  synfile="~/src/git/cadmium/input/syn/Arabidopsis_thaliana.vs.Arabidopsis_lyrata.syn",
  gfffile="~/src/git/cadmium/input/gff/Arabidopsis_lyrata.gff"
)
nstring <- LoadNString('~/src/git/cadmium/input/stat/nstrings.tab')[['Arabidopsis_lyrata']]
sflags <- summarize.flags(target$si)
@

\subsection{Mapping features between GFF given scaffold}

<<cache=TRUE>>=
origins <- initializeOrigins(query, target)
origins <- syntenicState(origins, sflags)
bittbl  <- getBitTable(origins)
fo.cds  <- analyzeTargetFeature(query, target, feature="CDS")
fo.mrna <- analyzeTargetFeature(query, target, feature="mRNA")
@

\begin{figure}[!ht]
    \centering
<<fig.width=6, fig.height=4, message=FALSE, cache=TRUE>>=
require(reshape2)
require(ggplot2)
m <- melt(
    bittbl[c('bit', 'non_orp_prop', 'orp_prop')],
    id.vars='bit',
    value.name='proportion',
    variable.name='class'
)
ggplot(m) +
    geom_point(aes(x=bit, y=proportion, color=class)) +
    geom_line(aes(x=bit, y=proportion, group=class)) +
    xlab('Synteny flags') +
    ylab('Proportion') +
    theme(
        axis.text.x = element_text(angle=270, hjust=0, vjust=1),
        legend.position=c(1, 1),
        legend.justification=c(1, 1)
    )
@
  \caption{\textbf{Synteny flags}. \textbf{(0)} the search interval is reliable;
    \textbf{(1)} the start edge is unreliable; \textbf{(2)} the stop edge is
    unreliable; \textbf{(3)} edges are unreliable, but there is overlap;
    \textbf{(4)} syntenically unreliable}.
  \label{fig:synflags}
\end{figure}



<<results='asis'>>=
require(xtable)
xtable(bittbl, caption='Nmumber of genes in each syntenic class.')
@



\subsection{Find queries that overlap target CDS}

\begin{figure}[!ht]
    \centering
<<fig.width=6, fig.height=4, message=FALSE, cache=TRUE>>=
require(ggplot2)
require(scales)
featureCountTable(fo.cds) %>%
  {data.frame(
    N.cds=as.numeric(row.names(.)),
    count=.$Count)
  } %>%
  ggplot +
    geom_point(aes(x=N.cds, y=count)) +
    xlab('Number of overlapped CDSs') +
    ylab('Number of search intervals (log2)') +
    scale_y_continuous(
      trans='log2',
      breaks=trans_breaks('log2', function(x) round(2^x))
    )
@
  \caption{Number of CDS's overlapped by a single search interval}.
\end{figure}

<<results='asis'>>=
require(xtable)
featureCountTable(fo.mrna) %>%
  xtable(caption='Number of target mRNAs overlapping one search interval')
@

<<>>=
n.orp.cds  <- fo.cds$orphan2target$query  %>% unique %>% length
n.orp.mrna <- fo.mrna$orphan2target$query %>% unique %>% length
n.orp <- query$orphans %>% unique %>% length
@
\begin{table}[!ht]
    \centering
    \label{tab:label}
    \begin{tabular}{l | l | l}
    feature & count & percent \\ 
    \hline
    CDS  & \Sexpr{n.orp.cds}  & \%\Sexpr{signif(100 * n.orp.cds  / n.orp, 2)} \\
    mRNA & \Sexpr{n.orp.mrna} & \%\Sexpr{signif(100 * n.orp.mrna / n.orp, 2)} \\
    \hline
    \end{tabular}
    \caption{Number of orphans with search intervals overlapping given features}
\end{table}



\subsection{Find queries overlapping N-strings}

<<>>=
# Extract the genomeInterval object of intervals of defined size
si <- target$si$target[target$si$target %>% size %>% is.na %>% not]

# Find overlaps between search intervals and N-strings
o.n <- interval_overlap(si, nstring)
# There should be one entry in o for each feature in gff
stopifnot(length(o.n) == nrow(si))


# List of genes where at least on search interval maps to a gap
ids <- which(lapply(o.n, length) > 0)
query2gap <- o.n[ids]
ids <- rep(ids, times=lapply(query2gap, length) %>% unlist)
stopifnot(length(ids) == query2gap %>% unlist %>% length)

query2gap <- query2gap %>%
  unlist %>% 
  {
    data.frame(
      query=target$si$query$seqid[ids],
      length=size(nstring[., ])
    )
  }

@

\subsection{Find protein matches}

<<>>=
origins$si.overlaps.cds  <- origins$seqid %in% (fo.cds$query2target$query %>% unique)
origins$si.overlaps.mRNA <- origins$seqid %in% (fo.mrna$query2target$query %>% unique)
origins$si.overlaps.gap  <- origins$seqid %in% query2gap$query

orp.origins <- subset(origins, orphan)

scr <- orp.origins$scrambled
cds <- orp.origins$si.overlaps.cds
rna <- orp.origins$si.overlaps.mRNA

rel <- orp.origins$bit == '10000'

label <- ifelse(scr, 'Scrambled', 'Unknown')
label <- ifelse(!(rna | scr), 'Possible-intergenic', label)
label <- ifelse(cds, 'Possible-genic', label)
label <- ifelse(rna & !cds, 'Possible-hitchhiker', label)
label <- ifelse(rel & !(rna | scr), 'intergenic', label)
label <- ifelse(rel & cds, 'Possible-genic+', label)
label <- ifelse(rel & rna & !cds, 'hitchhiker', label)

label %>% factor %>% summary

orp.origins$label <- label
@

<<>>=
qseqs <- orp.origins %>%
  subset(label %in% c('Possible-genic+', 'Possible-genic')) %$%
  seqid %>% as.character %>% unique %>%
  query$aa[.]

aln1 <- function(qname) {
  fo.cds$orphan2target %>%
    subset(query == qname) %$% target %>% target$aa[.] %>%
    {AA_aln(qseqs[qname], ., aln_type='local')}
}

aln <- names(qseqs) %>% lapply(aln1)
names(aln) <- names(qseqs)

genic.orp <- lapply(aln, score) %>%
  unlist %>% {.[.>40]} %>% names %>%
  sub(pattern='\\.(.).*', replacement='.\\1') %>% unique

genic.loci <- lapply(aln, score) %>%
  unlist %>% {.[.>40]} %>% names %>%
  sub(pattern='\\..*', replacement='') %>%
  unique
@

<<echo=TRUE>>=
genic.loci
@

<<echo=TRUE>>=
# Here is an example of one of the genic orphans
writePairwiseAlignments(aln[[genic.orp[1]]])
@


\subsection{Detect overprinting}

\subsection{Detect protein hit to ngORF}

\subsection{Find DNA hit}

\subsection{Final Classifications}

<<echo=TRUE>>=
label %>% factor %>% summary
@
