---
title: "Vignette Title"
author: "Vignette Author"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r}
library(fagin)
```

Fagin is a program for analyzing orphan genes. The first step of the analysis
is gathering the required data. Inputs are organized into a `config` object,
below are the default values:

```{r}
myConfig <- config()
```

There are many parameters that may be set, but the only ones we need to think
about for now are those under the input header. The command `config()` creates
a default config as an R S4 class. The values may be modified:

```{r, eval=FALSE}
myConfig <- config()
myConfig@input@gff_dir = "foo/gff"
```

Once the config is loaded with paths to the input data, you can run the full
fagin pipeline.

```{r, eval=FALSE}
run_fagin(myConfig)
```

This will take awhile to run, from a few minutes to a few hours, depending on
the size of your data. It may also require substantial memory, more than a
typical desktop will have for large genomes.

By default, all the results from the analysis are archived in the folder
`ARCHIVE` in the working directory. The following files are archived:

 1. `d1.Rda` - gathering and checking raw data

 2. `d2.Rda` - comparisons of target species to the focal species

 3. `d3.Rda` - building a feature table from the target/focal comparisons

 4. `d4.Rda` - labels and trees built from the feature table

 5. `d5.Rda` - results of superimposing labels on the phylogenetic tree

 6. `final_obj.Rda` -  the final rmonad object

These can be loaded with commands such as `load('ARCHIVE/d1.Rda')`.
