---
title: "Yeast Report"
author: "Zebulun Arendsee"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Yeast Report}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, echo=FALSE, message=FALSE, eval=FALSE}
library(fagin)
library(rmonad)
library(knitr)
library(magrittr)
# The default config uses the sample data for yeast
con <- config()
con@synder@offsets = c(1L,1L) # offsets for Satsuma
validate_config(con)
# species_name=con@input@focal_species
# plot(m, vertex.size=2, edge.arrow.size=0.2, vertex.label=NA)
d <- run_fagin(con)
missues(d) %>% kable(format='html') %>% write('ARCHIVE/issues.html')
mtabulate(d, code=TRUE) %>% kable(format='html') %>% write('ARCHIVE/table.html')
# Get a list of input summaries
summaries <- getSummaries(d)
# TODO: reconfigure exploration tools for rmonad (keep pure)
makeExcelSpreadsheet(con, filename="ARCHIVE/results.xlsx")
```

## Input data summary

```{r, echo=FALSE, eval=FALSE}
load('ARCHIVE/d1.Rda')
ss <- invertSummaries(d1)
```

```{r, fig.width=6, fig.height=4, echo=FALSE, message=FALSE, eval=FALSE}
plot(d1@tree)
```

```{r, echo=FALSE, result='asis', eval=FALSE}
speciesOrder <- getSpeciesPhylogeneticOrder(con)
knitr::kable(
  makeGenomeTable(ss, speciesOrder),
  caption="**Genomic statistics**. *scafs* - number of scaffolds in the genome. *bases* - total number of bases in the genome. *prots* - total number of protein coding models. *GC* - the percent of G and C nucleotides in the genome. *oddstart* - number of coding sequences that do not begin with M. *oddstop* - number of coding sequences that do not end with a stop (\\*). *p_N* - number of unknown nucleotides (N) in the genome. *n_stop* - number of coding sequences with an internal stop."
)
```

```{r, echo=FALSE, result='asis', eval=FALSE}
knitr::kable(
    makeSynmapTable(con, speciesOrder),
    caption="Lengths of homologous links in each synteny map."
)
```

## `synder` output

```{r, echo=FALSE, results='asis', message=FALSE, eval=FALSE}
knitr::kable(
    makeSynderTable(con, speciesOrder),
    caption="Search interval widths"
)
```

## Fagin output

```{r, fig.width=8, fig.height=6, echo=FALSE, eval=FALSE}
plotSecondaryLabels(con, speciesOrder)
```

```{r, fig.width=8, fig.height=4, echo=FALSE, eval=FALSE}
plotSecondaryLabels(con, speciesOrder, fill='species')
```

## Origin classifications

**TODO - fill this out**

## System info

**TODO - dump the collected session info here**
