---
title: "Yeast orphan analysis"
author: "Zebulun Arendsee"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Yeast orphan analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, echo=FALSE, message=FALSE}
# document()
library(rmonad)
library(ggplot2)
library(knitr)
library(dplyr)
con <- config()
# file.symlink("../inst/yeast", "INPUT")
# d <- run_fagin(con)
```

```{r, echo=FALSE}
# missues(d) %>% kable(format='html') %>% write('ARCHIVE/issues.html')
# mtabulate(d, code=TRUE) %>% kable(format='html') %>% write('ARCHIVE/table.html')
# es <- as.list(d)[sapply(d, function(x) x$has_error())]
# ws <- as.list(d)[sapply(d, function(x) x$has_warnings())]
```

```{r, echo=FALSE}
invertSummaries <- function(x){
  xs <- list(
    gff         = list(),
    dna         = list(),
    aa          = list(),
    trans       = list(),
    orfgff      = list(),
    orffaa      = list(),
    transorfgff = list(),
    transorffaa = list(),
    nstring     = list()
  )
  for(s in names(x@species)){
    xs$gff         = append(xs$gff         , x@species[[s]]@summaries@gff.summary         )
    xs$dna         = append(xs$dna         , x@species[[s]]@summaries@dna.summary         )
    xs$aa          = append(xs$aa          , x@species[[s]]@summaries@aa.summary          )
    xs$trans       = append(xs$trans       , x@species[[s]]@summaries@trans.summary       )
    xs$orfgff      = append(xs$orfgff      , x@species[[s]]@summaries@orfgff.summary      )
    xs$orffaa      = append(xs$orffaa      , x@species[[s]]@summaries@orffaa.summary      )
    xs$transorfgff = append(xs$transorfgff , x@species[[s]]@summaries@transorfgff.summary )
    xs$transorffaa = append(xs$transorffaa , x@species[[s]]@summaries@transorffaa.summary )
    xs$nstring     = append(xs$nstring     , x@species[[s]]@summaries@nstring.summary     )
  }
  lapply(xs, function(xx) { names(xx) <- names(x@species); xx } )
}
```

```{r, echo=FALSE}
load('ARCHIVE/d1.Rda')
ss <- invertSummaries(d1)
```

```{r, echo=FALSE}
load('ARCHIVE/d2.Rda')

number_of_chromosomes <- function(x){
    lapply(x$dna, function(s) nrow(s@table))
}

initial_protein_residue_counts <- function(x){
    lapply(x$aa, function(s) s@initial_residue)
}

final_protein_residue_counts <- function(x){
    lapply(x$aa, function(s) s@final_residue)
}

protein_has_internal_stop <- function(x){
    lapply(x$aa, function(s) s@has_internal_stop %>% sum)
}

genomic_composition <- function(x){
    x$dna %>% lapply(function(s) s@comp %>% colSums %>% { . / sum(.) })
}
```

```{r, fig.width=8, fig.height=6, echo=FALSE, message=FALSE}
require(ape)
plot(d1@tree)
speciesOrder <- d1@tree$tip.label
```

```{r, echo=FALSE, result='asis'}
nscaf <- number_of_chromosomes(ss)
initialRes <- initial_protein_residue_counts(ss)
finalRes <- final_protein_residue_counts(ss)
hasStop <- protein_has_internal_stop(ss)
genComp <- genomic_composition(ss)

data.frame(
    species = names(nscaf) %>% factor(levels=speciesOrder)
  , n_scaffolds = nscaf %>% unlist %>% unname
  , n_bases = sapply(ss$dna, function(x) x@table$length %>% sum)
  , n_prot = sapply(initialRes, sum) %>% unname
  , GC = sapply(genComp, function(x) ((x['G'] + x['C']) * 100) %>% round)
  , n_oddstart = sapply(initialRes, function(x) { sum(x) - x['M'] }) %>% unname
  , n_oddstop = sapply(finalRes, function(x) { sum(x) - x['*'] }) %>% unname
  , p_N = sapply(genComp, function(x) x['N'] %>% signif(3))
  , n_stop = hasStop %>% unlist %>% unname
) %>%
    dplyr::arrange(species) %>%
    knitr::kable()
```

```
    O1 = genic: known gene
    O2 = genic: unknown ORF on known mRNA
    O3 = genic: unknown ORF off known mRNA
    U1 = unknown: maps off the scaffold
    U2 = unknown: possible indel
    U3 = unknown: possible N-string
    U5 = unknown: syntenically scrambled
    U4 = -- RIP - there is no 4
    U6 = unknown: seriously unknown
    U7 = unknown: skipped for technical reasons
    N1 = non-genic: DNA match to CDS
    N2 = non-genic: DNA match to exon
    N3 = non-genic: DNA match to mRNA
    N4 = non-genic: No DNA match to any known gene
```

```{r, echo=FALSE}
# load the binary feature table
load('ARCHIVE/d3.Rda')
```

```{r, echo=FALSE}

plotSecondaryLabels <- function(con, fill='secondary'){
    load(file.path(con@archive, 'd4.Rda'))
    if(fill == 'secondary'){
        d4$labels %>%
            lapply(
                function(x) {
                    dplyr::group_by(x[-1], primary, secondary) %>% dplyr::count()
                }
            ) %>%
            dplyr::bind_rows(.id='species') %>%
            ggplot +
                geom_bar(aes(x=species, y=n, fill=secondary), position="dodge", stat="identity") +
                theme(
                    axis.text.x = element_text(angle=270, hjust=0, vjust=1)
                )
    } else {
        d4$labels %>%
            lapply(
                function(x) {
                    dplyr::group_by(x[-1], primary, secondary) %>% dplyr::count()
                }
            ) %>%
            dplyr::bind_rows(.id='species') %>%
            ggplot +
                geom_bar(aes(x=secondary, y=n, fill=species), position="dodge", stat="identity")
    }
}
```

```{r, fig.width=8, fig.height=6, echo=FALSE}
plotSecondaryLabels(con)
```

```{r, fig.width=8, fig.height=4, echo=FALSE}
plotSecondaryLabels(con, fill='species')
```

```{r, echo=FALSE}
load('ARCHIVE/d5.Rda')
d5$classSum
```

```{r, echo=FALSE}
load('ARCHIVE/final_obj.Rda')
```
