---
title: "Yeast Report"
author: "Zebulun Arendsee"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Yeast Report}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, echo=FALSE, message=FALSE}
document()
# library(fagin)
library(rmonad)
library(ggplot2)
library(knitr)
library(dplyr)
library(magrittr)
require(CNEr)
require(GenomicRanges)
con <- config()
file.symlink("../inst/yeast", "INPUT")
# file.symlink("/db/fagin-input/brassicaceae", "INPUT")
# con@input@focal_species = "Arabidopsis_thaliana"
# con@synder@offsets = c(0L,1L) # offsets for Satsuma
con@synder@offsets = c(1L,1L) # offsets for Satsuma
d <- run_fagin(con)
```

```{r, echo=FALSE}
as.list(d)[mtabulate(d)$error != 0] -> es
missues(d) %>% kable(format='html') %>% write('ARCHIVE/issues.html')
mtabulate(d, code=TRUE) %>% kable(format='html') %>% write('ARCHIVE/table.html')
es <- as.list(d)[sapply(d, function(x) x$has_error())]
ws <- as.list(d)[sapply(d, function(x) x$has_warnings())]
```

## Input data summaries

```{r, echo=FALSE}
load('ARCHIVE/d1.Rda')
ss <- invertSummaries(d1)
```

```{r, fig.width=8, fig.height=6, echo=FALSE, message=FALSE}
require(ape)
plot(d1@tree)
speciesOrder <- d1@tree$tip.label
```

```{r, echo=FALSE, result='asis'}
nscaf <- number_of_chromosomes(ss)
initialRes <- initial_protein_residue_counts(ss)
finalRes <- final_protein_residue_counts(ss)
hasStop <- protein_has_internal_stop(ss)
genComp <- genomic_composition(ss)
data.frame(
    species = names(nscaf) %>% factor(levels=speciesOrder)
  , n_scaffolds = nscaf %>% unlist %>% unname
  , n_bases = sapply(ss$dna, function(x) x@table$length %>% sum)
  , n_prot = sapply(initialRes, sum) %>% unname
  , GC = sapply(genComp, function(x) ((x['G'] + x['C']) * 100) %>% round)
  , n_oddstart = sapply(initialRes, function(x) { sum(x) - x['M'] }) %>% unname
  , n_oddstop = sapply(finalRes, function(x) { sum(x) - x['*'] }) %>% unname
  , p_N = sapply(genComp, function(x) x['N'] %>% signif(3))
  , n_stop = hasStop %>% unlist %>% unname
) %>%
    dplyr::arrange(species) %>%
    knitr::kable(caption="Genomic statistics")
```

```{r, echo=FALSE, result='asis'}
num_sum_as_table <- function(nd){
    data.frame(
        min    = sapply(nd, function(x) x@min)
      , q25    = sapply(nd, function(x) x@q25)
      , median = sapply(nd, function(x) x@median)
      , q75    = sapply(nd, function(x) x@q75)
      , max    = sapply(nd, function(x) x@max)
      , mean   = sapply(nd, function(x) x@mean)
      , sd     = sapply(nd, function(x) x@sd)
      , n      = sapply(nd, function(x) x@n)
    )
}

d1@synmaps %>%
    lapply(function(x) x@synmap.summary@width) %>%
    num_sum_as_table %>%
    knitr::kable(caption="Summary of the lengths of homologous links in each synteny map")
```

## `synder` search interval summaries

```{r, echo=FALSE, results='asis'}
load('ARCHIVE/d2.Rda')
lapply(d2, function(x){
    x$si %>%
        GenomicAlignments::second() %>%
        GenomicRanges::width() %>%
        fagin::summarize_numeric()
  }) %>%
    num_sum_as_table %>%
    knitr::kable(caption="Summary of search interval widths")
```

## Species-to-species classifications

```{r, echo=FALSE}
load('ARCHIVE/d3.Rda') # the binary feature table
```

```{r, fig.width=8, fig.height=6, echo=FALSE}
plotSecondaryLabels(con)
```

```{r, fig.width=8, fig.height=4, echo=FALSE}
plotSecondaryLabels(con, fill='species')
```

## Orphan origin stories 

```{r, echo=FALSE}
load('ARCHIVE/d5.Rda')
d5$classSum
```
