---
title: "Yeast orphan analysis"
author: "Zebulun Arendsee"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Yeast orphan analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r}
# document()
library(rmonad)
library(ggplot2)
library(knitr)
# con <- config()
# file.symlink("../inst/yeast", "INPUT")
# d <- run_fagin(con)
```

```{r}
# missues(d) %>% kable(format='html') %>% write('ARCHIVE/issues.html')
# mtabulate(d, code=TRUE) %>% kable(format='html') %>% write('ARCHIVE/table.html')
# es <- as.list(d)[sapply(d, function(x) x$has_error())]
# ws <- as.list(d)[sapply(d, function(x) x$has_warnings())]
```

```{r}
invertSummaries <- function(x){
  xs <- list(
    gff         = list(),
    dna         = list(),
    aa          = list(),
    trans       = list(),
    orfgff      = list(),
    orffaa      = list(),
    transorfgff = list(),
    transorffaa = list(),
    nstring     = list()
  )
  for(s in names(x@species)){
    xs$gff         = append(xs$gff         , x@species[[s]]@summaries@gff.summary         )
    xs$dna         = append(xs$dna         , x@species[[s]]@summaries@dna.summary         )
    xs$aa          = append(xs$aa          , x@species[[s]]@summaries@aa.summary          )
    xs$trans       = append(xs$trans       , x@species[[s]]@summaries@trans.summary       )
    xs$orfgff      = append(xs$orfgff      , x@species[[s]]@summaries@orfgff.summary      )
    xs$orffaa      = append(xs$orffaa      , x@species[[s]]@summaries@orffaa.summary      )
    xs$transorfgff = append(xs$transorfgff , x@species[[s]]@summaries@transorfgff.summary )
    xs$transorffaa = append(xs$transorffaa , x@species[[s]]@summaries@transorffaa.summary )
    xs$nstring     = append(xs$nstring     , x@species[[s]]@summaries@nstring.summary     )
  }
  lapply(xs, function(xx) { names(xx) <- names(x@species); xx } )
}
```

```{r}
load('ARCHIVE/d1.Rda')
ss <- invertSummaries(d1)
```

```{r}
ss$aa %>% lapply(function(x) x@initial_residue)
```

```{r}
ss$aa %>% lapply(function(x) x@final_residue)
```

```{r}
ss$aa %>% lapply(function(x) x@has_internal_stop %>% sum)
```

```{r}
ss$dna %>% lapply(function(x) x@table %>% nrow)
ss$dna %>% lapply(function(x) x@comp %>% colSums %>% { . / sum(.) })
```


```{r}
load('ARCHIVE/d2.Rda')
```

```{r, fig.width=8, fig.height=6}
require(ape)
plot(d1@tree)
```

```
    O1 = genic: known gene
    O2 = genic: unknown ORF on known mRNA
    O3 = genic: unknown ORF off known mRNA
    U1 = unknown: maps off the scaffold
    U2 = unknown: possible indel
    U3 = unknown: possible N-string
    U5 = unknown: syntenically scrambled
    U4 = -- RIP - there is no 4
    U6 = unknown: seriously unknown
    U7 = unknown: skipped for technical reasons
    N1 = non-genic: DNA match to CDS
    N2 = non-genic: DNA match to exon
    N3 = non-genic: DNA match to mRNA
    N4 = non-genic: No DNA match to any known gene
```

```{r}
# load the binary feature table
load('ARCHIVE/d3.Rda')
```

```{r, fig.width=8, fig.height=6}
load('ARCHIVE/d4.Rda')
d4$labels %>%
    lapply(
        function(x) {
            dplyr::group_by(x[-1], primary, secondary) %>% dplyr::count()
        }
    ) %>%
    dplyr::bind_rows(.id='species') %>%
    ggplot +
        geom_bar(aes(x=species, y=n, fill=secondary), position="dodge", stat="identity") +
        theme(
            axis.text.x = element_text(angle=270, hjust=0, vjust=1)
        )
```

```{r, fig.width=8, fig.height=4}
d4$labels %>%
    lapply(
        function(x) {
            dplyr::group_by(x[-1], primary, secondary) %>% dplyr::count()
        }
    ) %>%
    dplyr::bind_rows(.id='species') %>%
    ggplot +
        geom_bar(aes(x=secondary, y=n, fill=species), position="dodge", stat="identity")
```

```{r}
load('ARCHIVE/d5.Rda')
d5$classSum
```

```{r}
load('ARCHIVE/final_obj.Rda')
```
