\documentclass{article}
\usepackage{amsmath}
\usepackage{fullpage}
\usepackage{longtable}
\usepackage[section]{placeins}
\title{Genome Report}
\author{Zebulun Arendsee}
\begin{document}

\maketitle

<<require, echo=FALSE, message=FALSE>>=
  require(ggplot2)
  require(dplyr)
  require(ape)
  require(magrittr)
  require(scales)
@

\section{Species phylogeny}

<<view-tree, echo=FALSE>>=
tree <- read.tree('../input/tree')
plot(tree)
@

\section{Genomic statistics}

\begin{figure}
    \center
<<plot-scaffold-lengths, echo=FALSE, message=FALSE, cache=TRUE>>=
scaflen <- read.table('../input/stat/scaffold-lengths.tab', header=TRUE)
scaflen <- group_by(scaflen, species) %>%
    arrange(-length) %>%
    mutate(idx=1:length(scaffold)) %>%
    ungroup %>%
    droplevels
ggplot(scaflen) +
    geom_point(aes(x=idx, y=length), size=0.2) +
    scale_y_continuous(
        trans='log10',
        breaks=trans_breaks('log10', function(x) round(10^x))
    ) +
    facet_wrap(~species, scale='free_x')
@
    \caption{\textbf{Scaffold lengths}. Sorted by length on x-axis.}
\end{figure}

\begin{figure}
    \center
<<echo=FALSE, cache=TRUE, fig.width=7, fig.height=6>>=
scaflen$chrid <- factor(scaflen$idx)
scaflen_chr <- subset(scaflen, length > 10^7)
ggplot(scaflen_chr) +
    geom_path(aes(x=chrid, y=length, group=species)) +
    geom_point(aes(x=chrid, y=length, color=species), size=4) +
    xlab('Scaffold number (by size)') +
    ylab('Scaffold length (in DNA bases)') +
    theme(
        legend.position=c(1, 1),
        legend.justification=c(1, 1)
    )
@
    \caption{Lengths of scaffolds greater than 10 Mb. Sorted by size.}
\end{figure}

<<plot-nstring, echo=FALSE, cache=TRUE>>=
nstring <- read.table('../input/stat/nstrings.tab', header=TRUE)
nstring$length <- with(nstring, stop - start + 1)

d <- rename(scaflen_chr, scaffold_length = length) %>%
     merge(nstring, by=c('species', 'scaffold'))
d$chrid <- as.numeric(d$chrid)

nstring.g <- ggplot(d) +
    geom_segment(
        aes(
            y=chrid,
            yend=chrid,
            x=start,
            xend=stop),
        size=5) +
    geom_segment(
        aes(
            y=chrid,
            yend=chrid,
            x=1,
            xend=scaffold_length),
        size=0.2) +
    geom_text(
        aes(
            y=chrid + 0.2,
            x=scaffold_length,
            label=scaffold),
        hjust='left',
        family='sans',
        fontface='plain',
        size=2) +
    facet_wrap(~species, scale='free_y')
@
\begin{figure}[!ht]
    \centering
<<echo=FALSE>>=
nstring.g
@
    \caption{\textbf{Strings of unknowns}. Most assemblies have certain regions
        who lengths may be known but no sequence is available, these regions
        may be replaced with strings of consecutive N's. The vertical bars in
        this plot correspond to these regions, with width proportional to the
        number of N's. Only scaffolds with lengths greater than 10 Mb are shown.}
\end{figure}

<<plot-gc, echo=FALSE, cache=TRUE>>=
kbcomp <- read.table('../input/stat/kb-composition.tab', header=TRUE)
kbcomp <- subset(kbcomp, N == 0)
kbcomp$GC <- with(kbcomp, (C + G) / (A + C + G + T))
kbcomp.g <- ggplot(kbcomp) +
    geom_boxplot(aes(x=species, y=GC), width=0.2) +
    geom_violin(aes(x=species, y=GC), alpha=0.3) +
    ylab('G+C Proportion')
@
\begin{figure}[!ht]
  \centering
<<echo=FALSE>>=
kbcomp.g
@
    \caption{\textbf{GC content of 10000 random 1kb intervals}. Any interval with 1 or more N's was ignored.}
\end{figure}


\section{Proteomic statistics}

\end{document}
